### PR TITLE
fix: allow np.scalar to be used in Series.__getitem__

### LINKED ISSUE #1493: bug: can't use `Series.__getitem__` with numpy scalars
The consequence of this is:
```python
In [18]: import pandas as pd

In [19]: import narwhals as nw

In [20]: s = pd.Series([0,1,2])

In [21]: snw = nw.from_native(s, series_only=True)

In [22]: snw[snw.min()]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[22], line 1
----> 1 snw[snw.min()]

File ~/scratch/.venv/lib/python3.12/site-packages/narwhals/series.py:72, in Series.__getitem__(self, idx)
     70 if isinstance(idx, int):
     71     return self._compliant_series[idx]
---> 72 return self._from_compliant_series(self._compliant_series[idx])

File ~/scratch/.venv/lib/python3.12/site-packages/narwhals/_pandas_like/series.py:128, in PandasLikeSeries.__getitem__(self, idx)
    126 if isinstance(idx, int):
    127     return self._native_series.iloc[idx]
--> 128 return self._from_native_series(self._native_series.iloc[idx])

File ~/scratch/.venv/lib/python3.12/site-packages/narwhals/_pandas_like/series.py:139, in PandasLikeSeries._from_native_series(self, series)
    138 def _from_native_series(self, series: Any) -> Self:
--> 139     return self.__class__(
    140         series,
    141         implementation=self._implementation,
    142         backend_version=self._backend_version,
    143         dtypes=self._dtypes,
    144     )

File ~/scratch/.venv/lib/python3.12/site-packages/narwhals/_pandas_like/series.py:87, in PandasLikeSeries.__init__(self, native_series, implementation, backend_version, dtypes)
     79 def __init__(
     80     self,
     81     native_series: Any,
   (...)
     85     dtypes: DTypes,
     86 ) -> None:
---> 87     self._name = native_series.name
     88     self._native_series = native_series
     89     self._implementation = implementation

AttributeError: 'numpy.int64' object has no attribute 'name'
```

The place this needs changing is in the definition of `Series.__getitem__` for `_pandas_like`

### PR DESCRIPTION
**Pull request by thevro**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [X] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- closes #1493

## Checklist

- [X] Code follows style guide (ruff)
- [ ] Tests added
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below
```python
>>> import pandas as pd
>>> import narwhals as nw
>>> s = pd.Series([0,1,2])
>>> snw = nw.from_native(s, series_only=True)
>>> snw[snw.min()]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/git/narwhals/narwhals/series.py", line 72, in __getitem__
    return self._from_compliant_series(self._compliant_series[idx])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/git/narwhals/narwhals/series.py", line 243, in _from_compliant_series
    return self.__class__(
           ^^^^^^^^^^^^^^^
  File "/git/narwhals/narwhals/series.py", line 58, in __init__
    raise AssertionError(msg)
AssertionError: Expected Polars Series or an object which implements `__narwhals_series__`, got: <class 'numpy.int64'>.
```
Previous error was an `AttributeError` arising from `_pandas_like/series.py`.

### CODE REVIEW

**Review by MarcoGorelli — COMMENTED** [review]
thanks @thevro , can you add a test please?

File: narwhals/_pandas_like/series.py [hunk 1]
```diff
@@ -112,7 +113,8 @@ def __getitem__(self, idx: int) -> Any: ...
     def __getitem__(self, idx: slice | Sequence[int]) -> Self: ...
 
     def __getitem__(self, idx: int | slice | Sequence[int]) -> Any | Self:
-        if isinstance(idx, int):
+        np = get_numpy()
+        if np.isscalar(idx):
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/series.py:117]: Should this be a one liner to also check numpy is installed?

```suggestion
        if isisntance(idx, int) or ((np := get_numpy()) is not None and np.isscalar(idx)):
```

**Review by FBruzzesi — COMMENTED** [review]
Hey I left a tiny comment. I am concerned that if numpy is not installed then this could fail

**Comment by MarcoGorelli** [comment]
thanks @thevro , looks like there's some test failures

File: narwhals/_pandas_like/series.py (hunk 1 above)
Comment (thevro) [reply narwhals/_pandas_like/series.py:117]: This is pandas-specific code, so isn't numpy guaranteed to be installed? Still, there's no harm checking.
You have a typo: could you please change `isisntance` to `isinstance`? Then I will commit your suggestion.
If you know of a more general way to check for scalars, that might be preferable.

Comment (FBruzzesi) [reply narwhals/_pandas_like/series.py:117]: Yeah sorry, I realized that after the comment, apology for the inconvenience! I am resolving the comment

**Comment by thevro** [comment]
> thanks @thevro , looks like there's some test failures

The test that fails is the one I added to confirm the exception demonstrated in the PR description. Instead of raising an AttributeError in `narwhals/_pandas_like/series.py`, as described in #1493, we now raise an AssertionError in `narwhals/series.py`. Should I change the test so that it expects the new exception (which is still a bug), or shall we try to pass the test in its current form within this PR?

Here is the reason we still get an error:
Yesterday I initially applied the same approach of `isinstance`-comparing to `numbers.Number` instead of to `int` in _both_ files, and that caused code equivalent to the test I added today to pass. However, it occurred to me that pd.Series can be indexed by other scalars such as strings (scalar != non-Iterable, sadly), so a `np.isscalar` approach is more appropriate. But `narwhals/series.py` must work without numpy installed. I see two possible fixes:

1. if `__getitem__` in `narwhals/_pandas_like/series.py` receives a scalar, the code in `narwhals/series.py` shouldn't even be called.
2. `narwhals/series.py` should also check for scalars in general instead of just `int`s. We would need to find a way to do this without using numpy.

**Comment by MarcoGorelli** [comment]
I think you could just make a function in `narwhals/dependencies.py` called `is_numpy_scalar` in which you just do
```
return (np := get_numpy()) is not None and np.isscalar(obj)
```

Like this, it checks if `obj` is a NumPy scalar without importing NumPy. Like this, you can safely use it in `narwhals/series.py`

File: tests/series_only/scalar_index_test.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from __future__ import annotations
+
+import narwhals.stable.v1 as nw
+
+np = nw.dependencies.get_numpy()
+pd = nw.dependencies.get_pandas()
```
Comment (MarcoGorelli) [inline tests/series_only/scalar_index_test.py:6]: you can just import these directly here (though if you use `constructor_eager` they're probably not necessary)

**Review by MarcoGorelli — COMMENTED** [review]
thanks @thevro , this is close

File: narwhals/dependencies.py [hunk 3]
```diff
@@ -220,6 +220,11 @@ def is_numpy_array(arr: Any) -> TypeGuard[np.ndarray]:
     return (np := get_numpy()) is not None and isinstance(arr, np.ndarray)
 
 
+def is_numpy_scalar(scalar: Any) -> TypeGuard[np.generic]:
+    """Check whether `scalar` is a NumPy Array without importing NumPy."""
```
[the remark this answers is not in the document]
Comment (thevro) [reply narwhals/dependencies.py]: Thank you. See https://github.com/narwhals-dev/narwhals/pull/1515/commits/7a5aa74dd6ef3f63b765bfe14d9594c2bb9b1eab

File: tests/series_only/scalar_index_test.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_index() -> None:
+    s = pd.Series([0, 1, 2])
+    snw = nw.from_native(s, series_only=True)
+    assert snw[snw[0]] == np.int64(0)
```
[the remark this answers is not in the document]
Comment (thevro) [reply tests/series_only/scalar_index_test.py:12]: I didn't understand `constructor_eager` before, but having tried it out, it seems like it runs the test for different libraries. Correct?
The problem here is that, while `numpy.int64(0) == 0`, for some reason `<pyarrow.Int64Scalar: 0> != 0`. Not sure about other libraries...

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply tests/series_only/scalar_index_test.py:12]: thanks @thevro ! yeah `Series.__getitem__` should return a Python scalar for PyArrow. i'll just make a PR, then if you fetch and merge then the comparison should work

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply tests/series_only/scalar_index_test.py:12]: done https://github.com/narwhals-dev/narwhals/pull/1554

if you fetch upstream and then merge upstream/main then the `== 0` comparison should work

[the remark this answers is not in the document]
Comment (thevro) [reply tests/series_only/scalar_index_test.py:12]: And doing
```python
import pyarrow as pa
from pyarrow import compute as pc
pazero = pc.cast(0, pa.int64())
...
idx = snw[0]  # test indexing using builtin int
assert idx == 0 or idx == pazero # idx should have type numpy.int64 or pa.int64 
assert snw[idx] == 0 or idx == pazero # test indexing using third-party int
```
produces a test error:
```python
>       assert snw[idx] == 0 or idx == pazero # test indexing using numpy.int64

tests/series_only/scalar_index_test.py:19: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
narwhals/series.py:73: in __getitem__
    return self._from_compliant_series(self._compliant_series[idx])
narwhals/_arrow/series.py:378: in __getitem__
    return self._from_native_series(self._native_series[idx])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: 'pyarrow.lib.Int64Scalar' object cannot be interpreted as an integer

pyarrow/table.pxi:312: TypeError
```
I suppose this comes of handling only numpy scalars.

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply tests/series_only/scalar_index_test.py:12]: yup, fixed now, if you fetch and merge upstream/main then `== 0` should work

File: narwhals/series.py [hunk 5]
```diff
@@ -67,7 +68,7 @@ def __getitem__(self: Self, idx: int) -> Any: ...
     def __getitem__(self: Self, idx: slice | Sequence[int]) -> Self: ...
 
     def __getitem__(self: Self, idx: int | slice | Sequence[int]) -> Any | Self:
-        if isinstance(idx, int):
+        if isinstance(idx, int) or is_numpy_scalar(idx):
```
[the remark this answers is not in the document]
Comment (thevro) [reply narwhals/series.py:71]: As I understand it, we would ideally capture any sort of numpy scalar that can be used to index a pandas series. Can't strings, floats and datetime objects also be used?

File: tests/series_only/scalar_index_test.py (hunk 2 above)
Comment (thevro) [reply tests/series_only/scalar_index_test.py:6]: Thank you. I can confirm that `constructor_eager` works without them.

File: narwhals/series.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply narwhals/series.py:71]: no, we only allow positional indexing here

**Review by MarcoGorelli — APPROVED** [review]
thanks @thevro

**Pull request merged**